Port `MessageTrace` and `VMTracer` machinery from Hardhat

Xanewok opened the pull request:
Part of https://github.com/NomicFoundation/edr/issues/246

While this is not immediately at the top of the module dependency chain, the logic was pretty self-contained (so easier to port and verify) and it was helpful for me to port the important trace data types as it helped me understand how the information flows and what's used and how.

The biggest change is that this exposes a single `observe` method that can be used in HH to digest the traces directly on the Rust side, rather than 3 separate methods, without marshalling the data over FFI when fetching the stack trace (and possibly in the future if we remove the minimal node/vm EthereumJS events).

The companion branch is here https://github.com/NomicFoundation/hardhat/tree/refactor/port-vm-tracer; I still need to figure out how to streamline the process and at least prove that the PR works in a backwards-compatible fashion with Hardhat.

It might be helpful to see these files used when porting, when reviewing:
- https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-core/src/internal/hardhat-network/stack-traces/message-trace.ts
- https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-core/src/internal/hardhat-network/stack-traces/vm-tracer.ts
- https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-core/src/internal/hardhat-network/provider/vm/exit.ts

Xanewok commented on crates/edr_napi/src/trace/exit.rs:
```diff
@@ -0,0 +1,40 @@
+pub struct Exit(pub(crate) edr_solidity::exit::ExitCode);
+
+impl FromNapiValue for Exit {
+    unsafe fn from_napi_value(
```
I wasn't sure whether I should have `#[napi] pub enum ExitCode` in edr_napi since that's also directly used in the `edr_solidity::message_trace` module; I figured that it's okay for now to have the plain Rust enum and directly implement the N-API conversion on the JS `Exit` class (compatible with the existing Hardhat interface).

Wodann replied:
See my other comment, but I'd prefer using `#[napi] pub enum ExitCode` here to avoid the need for `unsafe` code.

Bugs in the N-API layer are particularly hard to debug, so if we can rely on auto-generated code, that's preferred.

Xanewok replied:
I followed this and have a dedicated `enum ExitCode` on the edr_solidity side with the Hardhat-specific `#[napi] pub enum ExitCode` in the edr_napi side, which does the relevant conversion.

Wodann commented on crates/edr_solidity/src/vm_tracer.rs:
```diff
@@ -0,0 +1,276 @@
+
+            trace = MessageTrace::Create(create_trace);
+        } else {
+            // TODO: Make this nicer and make sure the U160 logic/precompile comparison
+            // logic works
```
Todo. Are you planning to resolve this before merging?

If not, could you please create a GitHub issue for this and link to its URL here?

Wodann commented on crates/edr_solidity/src/vm_tracer.rs:
```diff
@@ -0,0 +1,276 @@
+            } else {
+                // if we enter here, then `to` is not None, therefore
+                // `code_address` and `code` should be Some
+                // TODO (if not true, then we need to return and set the error message)
```
Todo. Are you planning to resolve this before merging?

If not, could you please create a GitHub issue for this and link to its URL here?

Xanewok replied:
Thanks, this was a leftover TODO; I already did that part below.

Wodann commented on hardhat-tests/package.json:
```diff
@@ -53,10 +53,12 @@
     "eslint-plugin-prettier": "3.4.0",
     "ethereumjs-abi": "^0.6.8",
     "ethers": "^6.1.0",
+    "fs-extra": "11.2.0",
     "hardhat": "2.22.5",
     "mocha": "^10.0.0",
     "prettier": "2.4.1",
     "rimraf": "^3.0.2",
+    "semver": "7.6.0",
```
What are these dependencies used for?

Xanewok replied:
These were peer dependencies that ESLint warned me against. This is already in main as per https://github.com/NomicFoundation/edr/commit/048c0c0524ca785c0ea9b9f163887b3473903376

Wodann reviewed (changes requested):
Overall, I'm hesitant about committing these changes to `main`. Any commit in `main` will be released in the next NPM package that's cut.

Whereas most code is tested through the EDR and Hardhat test suites, this part of the code additionally requires testing of community plugins. This process is still manual and thus very time consuming.

If the intent of this PR is to be a stepping stone for the stack trace port, I'd propose merging this to a feature branch and continue the work there.

Xanewok commented:
I'll move this to draft for now, as it seems to be impossible to port some of the logic without exposing it into our public API via napi-rs.

One approach might be to collect the changes on the Rust side, however there is substantial amount of logic that has to touch the N-API surface due to how JS classes (i.e. codebase model) interact with each other, so we wouldn't get far with this approach.

Another alternative might be to conditionally compile the internal bits for now, however my intuition is that it might require quite a bit of work and setup for it to work correctly with napi-rs (its build system often leaves a bit to be desired). My changes also require modifying Hardhat to use the new API from the EDR package rather than in-tree logic in `hardhat-network/stack-traces` module; this might need setting up some patching mechanism as well while doing the migration.

I'm continuing work on my branch for now; I'm open to investigating whether the approach outlined above makes sense to facilitate incremental work but I'm also okay with progressing as is, i.e. working solely on my branch.

@fvictorio do you have any preference? Are you fine with me continuing the work and only opening the PR once we migrate almost all of it or would you prefer that we devote some time into supporting the partial migration without exposing it publicly via N-API?

Xanewok commented on crates/edr_solidity/src/exit.rs:
```diff
@@ -0,0 +1,89 @@
+/// Represents the exit code of the EVM. Naive Rust port of the `ExitCode` from
+/// Hardhat.
+#[derive(Clone, Copy, Debug)]
+pub enum ExitCode {
```
See https://github.com/NomicFoundation/edr/pull/531#discussion_r1688051843; I split them up and now we do not use `unsafe` anymore.

Xanewok commented on crates/edr_solidity/src/message_trace.rs:
```diff
@@ -0,0 +1,115 @@
+    pub value: U256,
+    /// Return data buffer.
+    pub return_data: Bytes,
+    /// EMV exit code.
```
Thanks! Fixed

Xanewok commented on crates/edr_solidity/src/vm_tracer.rs:
```diff
@@ -0,0 +1,276 @@
+
+impl VmTracer {
+    /// Creates a new [`VmTracer`].
+    pub fn new() -> Self {
```
Sure, done.

Xanewok commented:
I have addressed the feedback and changed the base of the PR to target the feature branch of https://github.com/NomicFoundation/edr/pull/545.

As done in the #545, this additionally patches the Hardhat dep with a commit from https://github.com/NomicFoundation/hardhat/tree/refactor/port-vm-tracer; once it's merged, I will update the main companion branch of https://github.com/NomicFoundation/hardhat/tree/refactor/port-solidity-stack-traces to include that as well.

In the meantime, I may change the patching mechanism to support general `git checkout` as discussed on Slack with @fvictorio but that is unrelated to the changes here for now.

Wodann reviewed (changes requested):
It's hard to judge the JS parts for me, but as long as the tests are passing that should be alright. It would still be good if Franco could double-check.

There are some comments left for the Rust part. Otherwise looks good

Xanewok commented on crates/edr_napi/src/trace/message_trace.rs:
```diff
@@ -0,0 +1,177 @@
+                return_data: Uint8Array::from(precompile.base.return_data.as_ref()),
+                exit: Exit(precompile.base.exit.into())
+                    .into_instance(env)
+                    .expect("napi glue to register the Exit class constructor"),
```
I know and was considering it but I decided against it, because IMO it'd obscure a bit what's actually happening here and the error can only occur if there's an internal bug in the napi-rs itself. If I recall correctly, the error can only happen if napi failed to successfully register the class constructors at init time; if that happens then we probably can't reason much about the glue code anymore.

On the flip side, I see that we're not using `#[napi(catch_unwind)]` so I'm happy to change at least so that we don't bring down the entire runtime if the napi glue code fails.

Xanewok commented on crates/edr_napi/src/trace/message_trace.rs:
```diff
@@ -0,0 +1,177 @@
+    pub bytecode: Option<Value>,
+    pub number_of_subtraces: u32,
+    // `CallMessageTrace`
+    pub calldata: Uint8Array,
```
That's what's used in the original `CallMessageTrace` and I wanted to stay as faithful to the original for now as possible to limit any unforeseen differences or subtle reliance/breakage: https://github.com/NomicFoundation/hardhat/blob/6c943bb1dfdf84c1250f9e1f9b0add8e9e04649b/packages/hardhat-core/src/internal/hardhat-network/stack-traces/message-trace.ts#L49

fvictorio replied:
While `Buffer` is technically a sub-class of `Uint8Array`, in practice they have some annoying, subtle differences. I don't remember what exactly, but I remember being bitten by it in the past. So it makes sense to keep using `Uint8Array` here, unless there's a good reason not to (other than consistency).

Xanewok commented on crates/edr_solidity/src/vm_tracer.rs:
```diff
@@ -0,0 +1,270 @@
+impl VmTracer {
+    /// Creates a new [`VmTracer`].
+    pub const fn new() -> Self {
+        // TODO: temporarily hardcoded to remove the need of using ethereumjs' common
```
I've ported this verbatim from JS. I'm not familiar with the code and the context around it but it seems like "max precompile number" should probably be shared in the EDR itself? Or would you rather we supply this information from JS until everything is ported over?

fvictorio replied:
Oh, this one has some history.

When we were using EthereumJS, we would do something like `common.precompiles.length` (I don't remember the exact API) to get the max precompile number. This was then used during tracing to figure out whether a call was a precompile call.

After we moved to EDR, I tried to remove as much EJS dependencies as possible. Some usages of `Common`, like this one, were trickier, so I just hardcoded a max value here. This is of course wrong: if you use an older hardfork that doesn't have the `0xA` precompile, and you make a call to that address, Hardhat will tell you it was a precompile call, when it wasn't. I think this is a fine, temporary compromise, because this only affects user messages, not execution.

Once we finish porting this to the EDR side, this should be improved so that this comes from the EDR EVM somehow.

Also notice that there is another problem with this: after RIP-7212, the idea that there is a max precompile number, and that every address below it is a precompile, is wrong. So we should improve that too.

@Xanewok do you mind opening an issue about this, and linking it in the comment?

Xanewok replied:
Opened https://github.com/NomicFoundation/edr/issues/565; feel free to reword it or add more context :+1:

Xanewok replied:
Linked to the issue in the TODO note as well in https://github.com/NomicFoundation/edr/pull/531/commits/b455234127468948e6511124e486c0cb428340cb.

Xanewok commented on crates/edr_napi/src/trace/exit.rs:
```diff
@@ -0,0 +1,86 @@
+#[napi]
+impl Exit {
+    #[napi(getter)]
+    pub fn kind(&self) -> u8 {
```
Yep, good catch! Updated

Xanewok commented on crates/edr_napi/src/trace/message_trace.rs:
```diff
@@ -0,0 +1,177 @@
+            Either4::A(EvmStep { pc: step.pc as u32 })
+        }
+        edr_solidity::message_trace::MessageTraceStep::Message(msg) => {
+            // Immediately drop the borrow lock as it may be
```
Good catch, thanks! Updated

Xanewok commented:
Forgot to mark it as ready for review, sorry about that.

fvictorio commented on hardhat-tests/test/internal/hardhat-network/stack-traces/test.ts:
```diff
@@ -529,7 +533,8 @@ async function runTest(
     compareConsoleLogs(logger.lines, tx.consoleLogs);
 
-    const vmTraceDecoder = (provider as any)._vmTraceDecoder as VmTraceDecoder;
+    // @ts-ignore - this is a private field but don't `as any` for go-to-def to work
+    const vmTraceDecoder = provider._vmTraceDecoder as VmTraceDecoder;
```
I'd rather not use `@ts-ignore` anywhere (we never do in the Hardhat codebase), since it's a slippery slope. For this particular pattern, do this instead:

```js
    // eslint-disable-next-line @typescript-eslint/dot-notation
    const vmTraceDecoder = provider["_vmTraceDecoder"] as VmTraceDecoder;
```

Basically: typescript lets you access private properties using `["foo"]` but not `.foo` (no idea why), but we have an eslint rule to prefer dot-notation, so we need to disable it for this line.

The `as VmTraceDecoder` is still needed, but it _shouldn't_. I don't know why; it seems like a typescript bug. I tried to reproduce it in the playground but couldn't, so let's just keep the casting.

Xanewok replied:
Sure, makes sense! I wonder if that may have been fixed in the newer TypeScript versions? I think we're still using Hardhat's, which is ^5.0.4.

fvictorio replied:
No, I tried updating typescript and got the same result :confused:

fvictorio commented:
With respect to the Hardhat code, you can do this:

```typescript
export { ExitCode } from "@nomicfoundation/edr";
```

instead of

```
import { ExitCode } from "@nomicfoundation/edr";
export { ExitCode };
```

It makes it a bit clearer that it's just a re-export. Not super-important, just FYI.

fvictorio reviewed (approved):
LGTM. Thanks for the explanation in the description, it was very useful.

Xanewok commented:
TIL: 
```typescript
export { ExitCode } from "@nomicfoundation/edr";
```

I'll adapt the code in subsequent patches if that's okay to use this shorter syntax, not to have to re-squash and re-apply the commits in the Hardhat patch :pray:
